File: Data Structures & Algorithms/mergeSort/submission-0.py
# Definition for a pair.
# class Pair:
#     def __init__(self, key: int, value: str):
#         self.key = key
#         self.value = value
class Solution:
    def mergeSort(self, pairs: List[Pair]) -> List[Pair]:
        return self.mergeSortHelper(pairs, 0, len(pairs)-1)


    def mergeSortHelper(self, pairs: List[Pair], start: int, end:int) -> List[Pair]:
        if end-start+1 <= 1:
            return pairs

        mid = (start+end)//2

        self.mergeSortHelper(pairs, start, mid)
        self.mergeSortHelper(pairs, mid+1, end)

        self.merge(pairs, start, mid, end)

        return pairs

    def merge(self, pairs: List[Pair], start:int, mid: int, end:int) -> None:
        left_arr = pairs[start: mid+1]
        right_arr = pairs[mid+1: end+1]

        i = 0
        j = 0
        k = start

        while i < len(left_arr) and j < len(right_arr):
            if left_arr[i].key <= right_arr[j].key:
                pairs[k] = left_arr[i]
                i+=1
            else:
                pairs[k] = right_arr[j]
                j+=1

            k+=1

        while i < len(left_arr):
            pairs[k] = left_arr[i]
            i+=1
            k+=1

        while j < len(right_arr):
            pairs[k] = right_arr[j]
            j+=1
            k+=1

File: Data Structures & Algorithms/quickSort/submission-0.cpp
// Definition for a Pair
// class Pair {
// public:
//     int key;
//     string value;
//
//     Pair(int key, string value) : key(key), value(value) {}
// };
class Solution {
public:
    void quickSortHelper(vector<Pair>& pairs, int start, int end) {
        if (end - start + 1 <= 1) {
            return;
        }

        Pair pivot = pairs[end];
        int left = start;

        for (int i = start; i < end; ++i) {
            if (pairs[i].key < pivot.key) {
                Pair temp = pairs[left];
                pairs[left] = pairs[i];
                pairs[i] = temp;
                
                left += 1;
            }
        }

        pairs[end] = pairs[left];
        pairs[left] = pivot;

        quickSortHelper(pairs, start, left-1);
        quickSortHelper(pairs, left+1, end);

    }
    
    vector<Pair> quickSort(vector<Pair>& pairs) {
        quickSortHelper(pairs, 0, pairs.size()-1);
        return pairs;
    }
};
